diff --git a/ant/Android.mk.template b/ant/Android.mk.template
index 9c75434..dc1424d 100644
--- a/ant/Android.mk.template
+++ b/ant/Android.mk.template
@@ -13,13 +13,25 @@
 # limitations under the License.
 #
 LOCAL_PATH := $(call my-dir)
+
+include $(CLEAR_VARS)
+LOCAL_MODULE := libcrypto
+LOCAL_SRC_FILES := {{root_prefix}}/android-$(TARGET_ARCH_ABI)/lib/libcrypto.a
+include $(PREBUILT_STATIC_LIBRARY)
+
+include $(CLEAR_VARS)
+LOCAL_MODULE := libsqlcipher
+LOCAL_SRC_FILES := {{root_prefix}}/android-$(TARGET_ARCH_ABI)/lib/libsqlcipher.a
+include $(PREBUILT_STATIC_LIBRARY)
+
 BASE := ../../../..

 include $(CLEAR_VARS)

 LOCAL_LDLIBS:=-llog
 LOCAL_MODULE    	:= sqlite4java-android
-LOCAL_SRC_FILES 	:= $(BASE)/sqlite/sqlite3.c $(BASE)/build/swig/sqlite_wrap.c $(BASE)/native/sqlite3_wrap_manual.c $(BASE)/native/intarray.c
-LOCAL_CFLAGS		:= -O2 -DNDEBUG -Dfdatasync=fsync -fno-omit-frame-pointer -fno-strict-aliasing -static-libgcc -I../../../sqlite -I../../../native -DSQLITE_ENABLE_COLUMN_METADATA -DSQLITE_ENABLE_FTS3 -DSQLITE_ENABLE_FTS3_PARENTHESIS -DSQLITE_ENABLE_MEMORY_MANAGEMENT -DSQLITE_ENABLE_STAT2 -DHAVE_READLINE=0 -DSQLITE_THREADSAFE=1 -DSQLITE_THREAD_OVERRIDE_LOCK=-1 -DTEMP_STORE=1  -DSQLITE_OMIT_DEPRECATED -DSQLITE_OS_UNIX=1 -DSQLITE_ENABLE_RTREE=1 -DHAVE_STRCHRNUL=0
+LOCAL_WHOLE_STATIC_LIBRARIES := libcrypto libsqlcipher
+LOCAL_SRC_FILES 	:= $(BASE)/build/swig/sqlite_wrap.c $(BASE)/native/sqlite3_wrap_manual.c $(BASE)/native/intarray.c
+LOCAL_CFLAGS		:= -O2 -DNDEBUG -fno-omit-frame-pointer -fno-strict-aliasing -static-libgcc -I../../../sqlite -I../../../native -fPIC

-include $(BUILD_SHARED_LIBRARY)
\ No newline at end of file
+include $(BUILD_SHARED_LIBRARY)
diff --git a/ant/Application.mk.template b/ant/Application.mk.template
index bb4d034..5eaff6d 100644
--- a/ant/Application.mk.template
+++ b/ant/Application.mk.template
@@ -1 +1 @@
-APP_ABI := armeabi-v7a armeabi x86
+APP_ABI := {{abi_list}}
diff --git a/ant/build.gant b/ant/build.gant
index 1e4e4ab..3265763 100644
--- a/ant/build.gant
+++ b/ant/build.gant
@@ -522,3 +523,20 @@ target(name: 'android') {
   ant.delete(dir: androidDir)
   ant.delete(dir: "$target/project")
 }
+
+target(name: 'sqlcipher-android') {
+  depends(['init', 'lib.swig'])
+  checkProp("ndk.home", "NDK_HOME")
+  def ndkHome = antProp("ndk.home")
+  def target = "$build/android"
+  ant.delete(dir: target)
+  ant.mkdir(dir: target)
+  def template = new File("$mod/ant/Android.mk.template")
+  if (!template.isFile()) ant.fail(message: "cannot open file $template")
+  ant.mkdir(dir: "$target/project")
+  ant.mkdir(dir: "$target/project/jni")
+  ant.copy(file: "Android.mk.template", tofile: "$target/project/jni/Android.mk")
+  ant.copy(file: "Application.mk.template", tofile: "$target/project/jni/Application.mk")
+
+  ant.exec(dir: "$target/project", command: "$ndkHome/ndk-build")
+}
